src/main/java/org/elasticsearch/search/aggregations/metrics/tophits/InternalTopHits.java

```diff
@@ -90,42 +90,86 @@ public SearchHits getHits() {
         return searchHits;
     }
 
+    private TopFieldDocs asTopFieldDocs(TopDocs topDocs) {
+        if (topDocs instanceof TopFieldDocs) {
+            return (TopFieldDocs) topDocs;
+        } else {
+            // TODO: fix whoever is doing this!
+            assert topDocs.scoreDocs.length == 0;
+            return new TopFieldDocs(0, Lucene.EMPTY_SCORE_DOCS, null, 0.0f);
+        }
+    }
+
     @Override
     public InternalAggregation reduce(ReduceContext reduceContext) {
         List<InternalAggregation> aggregations = reduceContext.aggregations();
-        TopDocs[] shardDocs = new TopDocs[aggregations.size()];
         InternalSearchHits[] shardHits = new InternalSearchHits[aggregations.size()];
+
+        // Find non-empty TopDocs:
         TopDocs topDocs = this.topDocs;
-        for (int i = 0; i < shardDocs.length; i++) {
-            InternalTopHits topHitsAgg = (InternalTopHits) aggregations.get(i);
-            shardDocs[i] = topHitsAgg.topDocs;
-            shardHits[i] = topHitsAgg.searchHits;
-            if (topDocs.scoreDocs.length == 0) {
-                topDocs = topHitsAgg.topDocs;
+        if (topDocs.scoreDocs.length == 0) {
+            for(InternalAggregation agg : aggregations) {
+                InternalTopHits hits = (InternalTopHits) agg;
+                if (hits.topDocs.scoreDocs.length > 0) {
+                    topDocs = hits.topDocs;
+                    break;
+                }
             }
         }
-        final Sort sort;
+
+        TopDocs reducedTopDocs;
         if (topDocs instanceof TopFieldDocs) {
-            sort = new Sort(((TopFieldDocs) topDocs).fields);
+            Sort sort = new Sort(((TopFieldDocs) topDocs).fields);
+
+            TopFieldDocs[] shardDocs = new TopFieldDocs[aggregations.size()];
+            for (int i = 0; i < shardDocs.length; i++) {
+                InternalTopHits topHitsAgg = (InternalTopHits) aggregations.get(i);
+                // TODO: topdocs.merge is this really safe?  our topDocs is a TopFieldDocs...
```

Comment:
@jpountz could you have a look at this one?  It made me nervous (not sure the stronger typing is safe).